### PR TITLE
Adding certificate login

### LINKED ISSUE #12: Add certificate logging option
Hello,

I've never worked on developing Actions so I don't know if it's possible but... I would like to improve this one by adding options to allow logging through a certificate path (not sure about this one) or encoded entries.

In a business context, I think it would be relevant to automatically login with credentials others than user / password.

What do you think? If it's fine, I can give it a try 🤗

### PR DESCRIPTION
**Pull request by michaelmaillot**
Closes #12 

Setting every input as optional (in order to be able to choose how to log in).

I was hesitating with adding an extra choice input, indicating authentication type setup (`password`, `certificate` and why not `secret`) but maybe the README file is enough to guide users.

Please let me know what you think regarding this PR, I'll adjust it from your feedback.

### CODE REVIEW

**Comment by garrytrinder** [comment]
Thank you @michaelmaillot we will review shortly 👍

File: README.md [hunk 1]
```diff
@@ -48,10 +66,50 @@ jobs:
 
     # CLI for Microsoft 365 login action
     - name: Login to tenant
-      uses: pnp/action-cli-login@v2.0.0
+      uses: pnp/action-cli-login@v3.0.0
```
Comment (waldekmastykarz) [inline README.md]: The currently published action is v2.0.1. If we want to use a different version here, shouldn't we increase the version number in package.json?

Since we're adding an extra option to login without breaking the existing functionality, I'd suggest that we move to a v2.1.0 instead of v3.0.0.

Comment (michaelmaillot) [reply README.md]: Oh I didn't see the version number on the package.json, I was only refering to the release notes in the README file.

The fact that all options are now optional (as they were required before) made me think about increasing major version. But I'm fine with increasing minor version.

**Comment by fladurner** [comment]
Would be great if this pull request could be reviewed and merged shortly, we also have the need of doing login using an app id.

File: README.md [hunk 2]
```diff
@@ -10,17 +10,35 @@ This GitHub Action (created using typescript) uses [CLI for Microsoft 365](https
 Create a workflow `.yml` file in your `.github/workflows` directory. An [example workflow](#example-workflow---cli-for-microsoft-365-login) is available below. For more information, reference the GitHub Help Documentation for [Creating a workflow file](https://help.github.com/en/articles/configuring-a-workflow#creating-a-workflow-file).
 
 ### Inputs
-- `ADMIN_USERNAME` : **Required** Username (email address of the admin)
-- `ADMIN_PASSWORD` : **Required** Password of the admin
+- `ADMIN_USERNAME` : Username (email address of the admin)
+- `PASSWORD` : Password for the admin or the certificate
```
Comment (waldekmastykarz) [inline README.md]: If we rename `ADMIN_PASSWORD` to `PASSWORD`, wouldn't that be a breaking change?

File: README.md [hunk 3]
```diff
@@ -48,10 +66,50 @@ jobs:
...
-        ADMIN_PASSWORD:  ${{ secrets.adminPassword }}
+        PASSWORD:  ${{ secrets.adminPassword }}
+    
+    ##
+    ## Code to deploy the package to tenant omitted
+    ##
+```
+
+### Example workflow - CLI for Microsoft 365 Login (certificate authentification)
+On every `push` build the code and then login to Office 365 before deploying, using certificate authentication.
```
Comment (waldekmastykarz) [inline README.md:90]: Let's add an empty line between the heading an the paragraph

File: README.md [hunk 4]
```diff
@@ -10,17 +10,35 @@ This GitHub Action (created using typescript) uses [CLI for Microsoft 365](https
...
+- `PASSWORD` : Password for the admin or the certificate
+- `CERTIFICATE_ENCODED` : Base64-encoded string with certificate private key
+- `AAD_APP_ID` : App ID of the Azure AD application to use for certificate authentication
+- `TENANT_ID` : ID or domain (for example "contoso.onmicrosoft.com") of the tenant from which accounts should be able to authenticate
+
+All inputs are optional. But depending of the authentication type chosen, following pair of inputs will be necessary:
+
+- `authType` is `password`: `ADMIN_USERNAME` and `PASSWORD` are required
+- `authType` is `certificate`: at least `CERTIFICATE_ENCODED` and `AAD_APP_ID` are required
+  - Depending on your tenant environment, `TENANT_ID` will be required
```
Comment (waldekmastykarz) [inline README.md]: _Depending on your tenant environment_

Can we clarify this?

File: README.md [hunk 5]
```diff
@@ -10,17 +10,35 @@ This GitHub Action (created using typescript) uses [CLI for Microsoft 365](https
...
-- `adminUsername` - store the admin user name in this (e.g. user@contoso.onmicrosoft.com)
-- `adminPassword` - store the password of that user in this.
+Since this action requires sensitive information such as user name, password and encoded certificate for example, it would be ideal to store them securely. We can achieve this in a GitHub repo by using [secrets](https://help.github.com/en/actions/automating-your-workflow-with-github-actions/creating-and-using-encrypted-secrets). So, click on `settings` tab in your repo and add:
+
+- 2 new secrets if `authType` is `password`:
+  - `adminUsername` - store the admin user name in this (e.g. user@contoso.onmicrosoft.com)
+  - `adminPassword` - store the password of that user in this.
+
+- 3 new secrets if `authType` is `certificate`:
+  - `AppID` - store the Azure AD application ID used for authentication
```
Comment (waldekmastykarz) [inline README.md]: Is there any reason why we suggest naming first 2 secrets starting with lowercase and these 3 secrets with uppercase?

File: dist/index.js [hunk 6]
```diff
@@ -2878,7 +2881,22 @@ function main() {
             cliMicrosoft365Path = yield (0, io_1.which)("m365", true);
             core.info("✅ Completed installing CLI for Microsoft 365.");
             core.info("ℹ️ Logging in to the tenant...");
-            yield executeCLIMicrosoft365Command(`login --authType password --userName ${username} --password ${password}`);
+            if (username && password) {
+                yield executeCLIMicrosoft365Command(`login --authType password --userName ${username} --password ${password}`);
```
Comment (waldekmastykarz) [inline dist/index.js]: As we're adding tenantId and appId, why not expose them to all login modes rather than just for the certificate?

**Review by waldekmastykarz — CHANGES_REQUESTED** [review]

File: README.md (hunk 2 above)
Comment (michaelmaillot) [reply README.md]: Sure yes. Would that be a problem?

File: README.md (hunk 5 above)
Comment (michaelmaillot) [reply README.md]: No reason at all, just inconstitency 🤷‍♂️

Let me adjust this.

I didn't find any stated naming convention except the Actions REST API [code sample](https://docs.github.com/en/rest/reference/actions#list-organization-secrets--code-samples) which returns full uppercase secrets. So maybe we could follow this?

File: dist/index.js (hunk 6 above)
Comment (michaelmaillot) [reply dist/index.js]: Right, thank you for pointing that out!

File: README.md (hunk 4 above)
Comment (michaelmaillot) [reply README.md]: Maybe I could copy / paste what is said on CLI for Microsoft 365 [login](https://pnp.github.io/cli-microsoft365/cmd/login/) documentation page, except the `CLIMICROSOFT365_TENANT` environment variable part? Like this:

```
ID of the tenant from which accounts should be able to authenticate. Use `common` or `organization` if the app is multitenant.
```

File: README.md (hunk 2 above)
Comment (waldekmastykarz) [reply README.md]: Unless necessary, we should avoid breaking changes as it causes extra work for our users to use the latest version. In this case, I'd suggest that we use two separate variables even if it means a bit more work for us.

File: README.md (hunk 4 above)
Comment (waldekmastykarz) [reply README.md]: That would be better indeed 👍

File: README.md (hunk 5 above)
Comment (waldekmastykarz) [reply README.md]: I'd say, let's align with the existing documentation so that we're not causing any breaking changes. We can always take a note of that to adjust in the next major version.

**Comment by garrytrinder** [comment]
> Would be great if this pull request could be reviewed and merged shortly, we also have the need of doing login using an app id.

Thank you for your comment @fladurner it's great to hear that this new feature will give immediate value for you.

We will be looking to review and, all being well, merge this weekend.

We ask for your patience, but we wanted you to know that this is on our radar 👍🏻

**Comment by fladurner** [comment]
> > Would be great if this pull request could be reviewed and merged shortly, we also have the need of doing login using an app id.
> 
> Thank you for your comment @fladurner it's great to hear that this new feature will give immediate value for you.
> 
> We will be looking to review and, all being well, merge this weekend.
> 
> We ask for your patience, but we wanted you to know that this is on our radar 👍🏻

That sounds great, thank you very much for your work!

**Comment by garrytrinder** [comment]
Thank you for working on the PR @michaelmaillot 

Following a review of this PR over the weekend, I have a concern with the lack of testing that we have on our GitHub Actions and that before we look to merge we should implement unit tests to cover the existing functionality of the action, so that we can have confidence in any future changes.

The lack of unit testing is on us (maintainers), I have raised an issue #15 which I have assigned to myself and I will look into as a priority, once complete we can revisit this PR and make the necessary changes.

**Comment by michaelmaillot** [comment]
Great idea, thank you. Feel free to reach out if you need any help (testing or anything else) 😉

**Comment by garrytrinder** [comment]
I've just merged #16 so we can now look to progress this PR.

@michaelmaillot I've made some changes to the codebase to make it much more testable therefore there will be some updates that need to be made, also note that the `master` branch has been renamed to `main`, to update your local clone execute the below git command.

```sh
git branch -m master main
git fetch origin
git branch -u origin/main main
git remote set-head origin -a
```

**Comment by fladurner** [comment]
Hi everyone, is there any chance that this pull request will be merged within the next days/weeks? Thank you all for your work!

**Comment by michaelmaillot** [comment]
Hi @fladurner I'm working on it, hope to bring something up in the near future 🙂

**Comment by garrytrinder** [comment]
Just checking, are you still working on this @michaelmaillot or is it now ready for review?

**Comment by michaelmaillot** [comment]
Ready for review!

Had to make a couple of tests first, before clicking on the well-known button 😉

**Comment by garrytrinder** [comment]
Awesome 👏🏻 

I'll take a look shortly, let's get this shipped 🚀

File: action.yml [hunk 7]
```diff
@@ -2,14 +2,26 @@ name: "CLI for Microsoft 365 Login"
...
+    description: "Base64-encoded string with certificate private key"
+    required: false
+  CERTIFICATE_PASSWORD:
+    description: "Password for the certificate"
+    required: false
+  AAD_APP_ID:
+    description: "App ID of the Azure AD application to use for certificate authentication"
+    required: false
+  TENANT_ID:
+    description: "ID of the tenant from which accounts should be able to authenticate"
```
Comment (garrytrinder) [inline action.yml]: Let's align this with the option name and description from `login` command for consistency as it can accept a tenant ID or name.

```suggestion
  TENANT:
    description: "ID of the tenant from which accounts should be able to authenticate. Use common or organization if the app is multitenant. If not specified, use the tenant specified in the 'CLIMICROSOFT365_TENANT' environment variable. If the environment variable is not defined, use 'common' as the tenant identifier"
```

File: action.yml [hunk 8]
```diff
@@ -2,14 +2,26 @@ name: "CLI for Microsoft 365 Login"
...
-    required: true
+    required: false
+  CERTIFICATE_ENCODED:
+    description: "Base64-encoded string with certificate private key"
+    required: false
+  CERTIFICATE_PASSWORD:
+    description: "Password for the certificate"
+    required: false
+  AAD_APP_ID:
+    description: "App ID of the Azure AD application to use for certificate authentication"
```
Comment (garrytrinder) [inline action.yml]: Let's align this with the option name and description from `login` command for consistency.

```suggestion
  APP_ID:
    description: "App ID of the Azure AD application to use for authentication. If not specified, use the app specified in the 'CLIMICROSOFT365_AADAPPID' environment variable. If the environment variable is not defined, use the multitenant PnP Management Shell app"
```

File: action.yml [hunk 9]
```diff
@@ -2,14 +2,26 @@ name: "CLI for Microsoft 365 Login"
 description: "Login to tenant using CLI for Microsoft 365"
 inputs:
   ADMIN_USERNAME:
-    description: "Username (upn) of the admin"
-    required: true
+    description: "Username (email address) of the admin"
```
Comment (garrytrinder) [inline action.yml]: Let's change this back to what is was as UPN is more appropriate than email address for authentication.

```suggestion
    description: "Username (upn) of the admin"
```

File: README.md [hunk 10]
```diff
@@ -58,13 +82,61 @@ jobs:
...
+    ## Code to get the package omitted
+    ##
+
+    # CLI for Microsoft 365 login action
+    - name: Login to tenant
+      uses: pnp/action-cli-login@v2.1.0
+      with:
+        AAD_APP_ID:  ${{ secrets.appID }}
+        CERTIFICATE_ENCODED: ${{ secrets.CERTIFICATE_ENCODED }}
+        CERTIFICATE_PASSWORD:  ${{ secrets.CERTIFICATE_PASSWORD }}
```
Comment (garrytrinder) [inline README.md]: Let's also include the `tenant` option here in the the example as it is required for certificate authentication.

```suggestion
      with:
          TENANT: ${{secrets.TENANT }}
          APP_ID:  ${{ secrets.APP_ID }}
          CERTIFICATE_ENCODED: ${{ secrets.CERTIFICATE_ENCODED }}
          CERTIFICATE_PASSWORD:  ${{ secrets.CERTIFICATE_PASSWORD }}
```

File: README.md [hunk 11]
```diff
@@ -7,20 +8,43 @@ This GitHub Action (created using typescript) uses [CLI for Microsoft 365](https
...
+Since this action requires sensitive information such as user name, password and encoded certificate for example, it would be ideal to store them securely. We can achieve this in a GitHub repo by using [secrets](https://help.github.com/en/actions/automating-your-workflow-with-github-actions/creating-and-using-encrypted-secrets). So, click on `settings` tab in your repo and add:
+
+- 2 new secrets if `authType` is `password`:
+  - `ADMIN_USERNAME` - store the admin user name in this (e.g. user@contoso.onmicrosoft.com)
+  - `ADMIN_PASSWORD` - store the password of that user in this.
+
+- 3 new secrets if `authType` is `certificate`:
+  - `AAD_APP_ID` - store the Azure AD application ID used for authentication
+  - `CERTIFICATE_ENCODED` - store the Base64-encoded string of the certificate stored in the Azure AD application
+  - `CERTIFICATE_PASSWORD` - store the certificate password
```
Comment (garrytrinder) [inline README.md]: As `APP_ID` can also be used with password login, let's not suggest that it is just specific to certificate authentication.

Let's create a we have a section that calls out `APP_ID` and `TENANT` separately.

```suggestion
- 2 new secrets if `authType` is `certificate`:
  - `CERTIFICATE_ENCODED` - store the Base64-encoded string of the certificate stored in the Azure AD application
  - `CERTIFICATE_PASSWORD` - store the certificate password

- 2 new secrets if using a custom Azure AD identity:
  - `APP_ID` - store App ID of the Azure AD application to use for authentication
  - `TENANT` - store the ID of the tenant from which accounts should be able to authenticate
 
```

File: README.md [hunk 12]
```diff
@@ -7,20 +8,43 @@ This GitHub Action (created using typescript) uses [CLI for Microsoft 365](https
...
 
 ### Inputs
-- `ADMIN_USERNAME` : **Required** Username (email address of the admin)
-- `ADMIN_PASSWORD` : **Required** Password of the admin
+
+- `ADMIN_USERNAME` : Username (email address of the admin)
+- `ADMIN_PASSWORD` : Password of the admin
+- `CERTIFICATE_ENCODED` : Base64-encoded string with certificate private key
+- `CERTIFICATE_PASSWORD` : Password for the certificate
+- `AAD_APP_ID` : App ID of the Azure AD application to use for certificate authentication
```
Comment (garrytrinder) [inline README.md]: Let's align this with the option name and description from `login` command for consistency.

```suggestion
- `APP_ID`: App ID of the Azure AD application to use for authentication. If not specified, use the app specified in the `CLIMICROSOFT365_AADAPPID` environment variable. If the environment variable is not defined, use the multitenant PnP Management Shell app
```

File: README.md [hunk 13]
```diff
@@ -7,20 +8,43 @@ This GitHub Action (created using typescript) uses [CLI for Microsoft 365](https
...
+- `TENANT_ID` : ID or domain (for example "contoso.onmicrosoft.com") of the tenant from which accounts should be able to authenticate
```
Comment (garrytrinder) [inline README.md]: Let's align this with the option name and description from `login` command for consistency.

```suggestion
- `TENANT`: ID of the tenant from which accounts should be able to authenticate. Use `common` or `organization` if the app is multitenant. If not specified, use the tenant specified in the `CLIMICROSOFT365_TENANT` environment variable. If the environment variable is not defined, use `common` as the tenant identifier
```

File: README.md [hunk 14]
```diff
@@ -7,20 +8,43 @@ This GitHub Action (created using typescript) uses [CLI for Microsoft 365](https
...
+
+All inputs are optional. But depending of the authentication type chosen, following pair of inputs will be necessary:
+
+- `authType` is `password`: `ADMIN_USERNAME` and `ADMIN_PASSWORD` are required
+- `authType` is `certificate`: at least `CERTIFICATE_ENCODED` and `AAD_APP_ID` are required
```
Comment (garrytrinder) [inline README.md]:
```suggestion
- `authType` is `certificate`: at least `CERTIFICATE_ENCODED` and `APP_ID` are required
```

File: README.md [hunk 15]
```diff
@@ -7,20 +8,43 @@ This GitHub Action (created using typescript) uses [CLI for Microsoft 365](https
...
+  - `TENANT_ID` will be required. It can be the ID of the tenant from which accounts should be able to authenticate. Use `common` or `organization` if the app is multitenant
```
Comment (garrytrinder) [inline README.md]:
```suggestion
  - `TENANT` will be required. ID of the tenant from which accounts should be able to authenticate. Use `common` or `organization` if the app is multitenant
```

File: README.md [hunk 16]
```diff
@@ -7,20 +8,43 @@ This GitHub Action (created using typescript) uses [CLI for Microsoft 365](https
...
 ## Usage
 ### Pre-requisites
+
 Create a workflow `.yml` file in your `.github/workflows` directory. An [example workflow](#example-workflow---cli-for-microsoft-365-login) is available below. For more information, reference the GitHub Help Documentation for [Creating a workflow file](https://help.github.com/en/articles/configuring-a-workflow#creating-a-workflow-file).
 
 ### Inputs
-- `ADMIN_USERNAME` : **Required** Username (email address of the admin)
-- `ADMIN_PASSWORD` : **Required** Password of the admin
+
+- `ADMIN_USERNAME` : Username (email address of the admin)
```
Comment (garrytrinder) [inline README.md]: Lets change this to align with the action input description

```suggestion
- `ADMIN_USERNAME` : Username (upn) of the admin
```

**Review by garrytrinder — CHANGES_REQUESTED** [review]
Great work @michaelmaillot just a few changes to make before we merge this in.

TL;DR;

- Lets update `AAD_APP_ID` and `TENANT_ID` inputs to `APP_ID` and `TENANT` respectively so that the action is consistent with the CLI `login` command
- `APP_ID` and `TENANT` can be used with both `password` and `certificate` login types

**Comment by michaelmaillot** [comment]
Thank you for your review, sounds better when keeping consistency with the CLI options indeed 🙂

**Comment by garrytrinder** [comment]
I've picked this up and will test further later this evening.

**Comment by garrytrinder** [comment]
Merged manually 🚀

**Comment by garrytrinder** [comment]
@fladurner certificate login is now possible in the latest release of the action, https://github.com/marketplace/actions/cli-for-microsoft-365-login thank you for your patience whilst we merged this PR ❤️

**Comment by fladurner** [comment]
> @fladurner certificate login is now possible in the latest release of the action, https://github.com/marketplace/actions/cli-for-microsoft-365-login thank you for your patience whilst we merged this PR ❤️

That's great, thank you so much for your great work!
